File: lmf2txt.py
#!/usr/bin/env python3

from functools import partial
from argparse import ArgumentParser

from lmfpy import LMFReader


parser = ArgumentParser()
parser.add_argument("filename", help="LMF data file")
parser.add_argument("-f", "--full",
                    help="if omitted only the header and the first 100 events are read",
                    action='store_true')
parser.add_argument("-n", "--ns",
                    help="convert values to nanoseconds",
                    action='store_true')
parser.add_argument("-o", "--output",
                    help="output, if not print it out",
                    type=str, default="")
args = parser.parse_args()


def main(reader: LMFReader, limit=None, k=None, out=None):
    if limit is None:
        limit = len(reader)
    if k is None:
        k = 1
    if out is None:
        out = partial(print, end="")

    out("               Filename: {}\n".format(reader.recorded_at))
    out("                Version: {}\n".format(reader.version))
    out("                Comment: {}\n".format(reader.comment))
    out("     Number of Channels: {}\n".format(reader.nchannels))
    out(" Maximum Number of Hits: {}\n".format(reader.max_nhits))
    out("  Number of Coordinates: {}\n".format(reader.ncoordinates))
    out("       Number of Events: {}\n".format(len(reader)))
    out("    TDC Resolution (ns): {:.6f}\n".format(reader.to_nanosec))
    out("             Start Time: {}\n".format(reader.time_fr))
    out("              Stop Time: {}\n".format(reader.time_to))

    for i, event in zip(range(limit), reader):
        if not i < limit:
            break
        out("########################\n")
        out("# {}\n".format(event.event))
        out("########################\n")
        out("                 At (s): {: 12.3f}\n".format(event.timestamp))
        for ch, n in enumerate(event.nhits):
            fr = sum(event.nhits[:ch])
            hits = ", ".join("{: 12.3f}".format(k*f) for f in event.hits[fr:fr+n])
            out("Channel #{:03d} {:3d} Hit(s): {}\n".format(ch, n, hits))


reader = LMFReader(args.filename)
limit = 100 if not args.full else None
k = None if not args.ns else reader.to_nanosec
if args.output != "":
    with open(args.output, 'w') as f:
        main(reader, limit=limit, k=k, out=f.write)
else:
    main(reader, limit=limit, k=k)

File: lmf2txt.cpp
#include <iostream>
#include <iomanip>

#include "lmfpy/LMFReader.h"


using namespace std;
using namespace lmfpy;

int main(int argc, char* argv[]) {
    if (argc == 1) {
        cout << "syntax: lmf2txt filename [--full] [--ns]" << endl;
        cout << "        if --full is omitted then only the header" << endl;
        cout << "        and the first 100 events will be written." << endl;
        cout << "        --ns converts the values to nanoseconds." << endl;
    }
    if (argc > 4) {
        cout << "wrong number of arguments." << endl;
        return 0;
    }

    string filename = argv[1];
    bool is_full = false, is_ns = false;
    if (argc >= 3)	{
        if (strcmp(argv[2], "--full") == 0) is_full = true;
        if (strcmp(argv[2], "--ns") == 0) is_ns = true;
    }
    if (argc == 4)	{
        if (strcmp(argv[3], "--full") == 0) is_full = true;
        if (strcmp(argv[3], "--ns") == 0) is_ns = true;
    }

    auto afile = make_shared<LMFReader>(filename);
    auto time_fr = chrono::system_clock::to_time_t(afile->time_fr());
    auto time_to = chrono::system_clock::to_time_t(afile->time_to());
    cout << "               Filename: " << afile->recorded_at() << endl;
    cout << "                Version: " << afile->version() << endl;
    cout << "                Comment: " << afile->comment() << endl;
    cout << "     Number of Channels: " << afile->nchannels() << endl;
    cout << " Maximum Number of Hits: " << afile->max_nhits() << endl;
    cout << "  Number of Coordinates: " << afile->ncoordinates() << endl;
    cout << "       Number of Events: " << afile->end() << endl;
    cout << "    TDC Resolution (ns): " << afile->to_nanosec() << endl;
    cout << "             Start Time: " << ctime(&time_fr);
    cout << "              Stop Time: " << ctime(&time_to);

    auto limit = is_full ? afile->end() : 100;
    auto i = 0;
    auto iter = afile->begin();
    for (; iter != afile->end(); ++i) {
        if (i >= limit) break;
        auto event = *iter;
        cout << "########################" << endl;
        cout << "# " << event.event << endl;
        cout << "########################" << endl;
        cout << "                 At (s): " << setw(12) << event.timestamp << endl;
        auto ch = 0, k = 0;
        auto n = event.nhits.begin();
        for (; n != event.nhits.end(); ++ch, ++k, ++n) {
            cout << "Channel #" << setw(3) << ch << " " << setw(3) << *n << " Hit(s): ";
            for (auto kk = 0; kk < *n; ++k, ++kk) {
                cout << setw(12) << event.hits[k] << ", ";
            }
            cout << endl;
        }
    }
    return 0;
}
